utils/zerovec/src/varzerovec/components.rs

```diff
@@ -0,0 +1,243 @@
+            return None;
+        }
+        Some(unsafe { self.get_unchecked(idx) })
+    }
+
+    /// Get the idx'th element out of this slice. Does not bounds check.
+    ///
+    /// Safety:
+    /// - `idx` must be in bounds (`idx < self.len()`)
+    #[inline]
+    unsafe fn get_unchecked(self, idx: usize) -> &'a T::VarULE {
+        let start = usizeify(*self.indices.get_unchecked(idx));
+        let end = if idx + 1 == self.len() {
+            self.things.len()
+        } else {
+            usizeify(*self.indices.get_unchecked(idx + 1))
+        };
+        debug_assert!(start <= end);
+        let things_slice = self.things.get_unchecked(start..end);
+        T::VarULE::from_byte_slice_unchecked(things_slice)
+    }
+
+    /// Create an iterator over the Ts contained in SliceComponents, checking internal invariants:
+    ///
+    /// - `indices[i]..indices[i+1]` must index into a valid section of
+    ///   `things`, such that it parses to a `T::VarULE`
+    /// - `indices[len - 1]..things.len()` must index into a valid section of
+    ///   `things`, such that it parses to a `T::VarULE`
+    /// - `indices` is monotonically increasing
+    ///
+    /// This method is NOT allowed to call any other methods on SliceComponents since all other methods
+    /// assume that the slice has been passed through iter_checked
+    #[inline]
+    fn iter_checked(self) -> impl Iterator<Item = Result<&'a T::VarULE, ParseErrorFor<T>>> {
+        let last = iter::from_fn(move || {
+            if self.len() > 0 {
+                let start = usizeify(self.indices[self.len() - 1]);
+                let end = self.things.len();
+                Some(
+                    self.things
+                        .get(start..end)
+                        .ok_or(VarZeroVecError::FormatError),
+                )
+            } else {
+                None
+            }
+        })
+        .take(1);
+        self.indices
+            .windows(2)
+            .map(move |win| {
+                let start = usizeify(win[0]);
+                let end = usizeify(win[1]);
+                // the .get() here automatically verifies that end>=start
+                self.things
+                    .get(start..end)
+                    .ok_or(VarZeroVecError::FormatError)
+            })
+            .chain(last)
+            .map(|s| s.and_then(|s| T::VarULE::parse_byte_slice(s).map_err(|e| e.into())))
+    }
+
+    /// Create an iterator over the Ts contained in SliceComponents
+    #[inline]
+    pub fn iter(self) -> impl Iterator<Item = &'a T::VarULE> {
+        let last = iter::from_fn(move || {
+            if self.len() > 0 {
+                let start = usizeify(self.indices[self.len() - 1]);
+                let end = self.things.len();
+                Some(unsafe { self.things.get_unchecked(start..end) })
+            } else {
+                None
+            }
+        })
+        .take(1);
+        self.indices
+            .windows(2)
```

Comment:
Suggestion (optional): Same here, with adding the indices to the iterator rather than the slice.